src/main/cpp/src/cast_string_to_float.cu

```diff
@@ -0,0 +1,730 @@
+      // strip trailing f if it exists
+      // f is a valid character at the end of a float string
+      auto const f_mask = __ballot_sync(
+        0xffffffff, (warp_lane == 0 && (c == 'F' || c == 'f' || c == 'd' || c == 'D')));
+      if (f_mask > 0) {
+        c = __shfl_down_sync(0xffffffff, c, 1);
+        bpos++;
+      }
+    }
+
+    // nothing trailing
+    if (blen - bpos == 0) { return; }
+
+    // strip any whitespace
+    remove_leading_whitespace();
+
+    // invalid characters in string
+    if (blen - bpos > 0) {
+      valid  = false;
+      except = true;
+    }
+  }
+
+  // sets validity bits, updates outgoing validity count for the block and potentially sets the
+  // outgoing ansi_except field
+  __device__ void compute_validity(bool const valid, bool const except = false)
+  {
+    // compute null count for the block. each warp processes one string, so lane 0
+    // from each warp contributes 1 bit of validity
+    size_type const block_valid_count =
+      cudf::detail::single_lane_block_sum_reduce<block_size, 0>(valid ? 1 : 0);
+    // 0th thread in each block updates the validity count and (optionally) the ansi_except flag
+    if (threadIdx.x == 0) {
+      atomicAdd(valid_count, block_valid_count);
+
+      if (ansi_except && except) { atomicMax(ansi_except, num_rows - row); }
+    }
+
+    // 0th thread in each warp updates the validity
+    size_type const row_id = warp_id;
+    if (threadIdx.x % 32 == 0 && valid) {
+      // uses atomics
+      cudf::set_bit(validity, row_id);
+    }
+  }
+
+  T* out;
+  bitmask_type* validity;
+  int32_t* ansi_except;
+  size_type* valid_count;
+  const char* const chars;
+  offset_type const* offsets;
+  size_type const tid;
+  size_type const warp_id;
+  size_type const row;
+  size_type const warp_lane;
+  size_type const row_start;
+  size_type const len;
+  size_type const num_rows;
+  uint64_t const* const ipow;
+  bitmask_type const* incoming_null_mask;
+
+  // shared/modified by the various parsing functions
+  size_type bstart;  // batch start within the entire string
+  size_type bpos;    // position with current batch
+  size_type blen;    // batch length;
+  char c;            // current character
+  bool valid  = true;
+  bool except = false;
+};
+
+template <typename T, size_type block_size>
+__global__ void string_to_float_kernel(T* out,
+                                       bitmask_type* validity,
+                                       int32_t* ansi_except,
+                                       size_type* valid_count,
+                                       const char* const chars,
+                                       offset_type const* offsets,
+                                       bitmask_type const* incoming_null_mask,
+                                       size_type const num_rows)
+{
+  size_type const tid     = threadIdx.x + (blockDim.x * blockIdx.x);
+  size_type const warp_id = tid / 32;
+  size_type const row     = warp_id;
```

Comment:
They seem redundant: The `string_to_float` functor later will compute them.

Reply:
Found a few of these and removed them. Thanks!